libraries/Microsoft.Bot.Builder.AI.TriggerTrees/Clause.cs

```diff
@@ -0,0 +1,305 @@
+                foreach (var shortPredicate in shorter.Predicates)
+                {
+                    var shorterRel = RelationshipType.Incomparable;
+                    if (!IsIgnore(shortPredicate))
+                    {
+                        foreach (var longPredicate in longer.Predicates)
+                        {
+                            shorterRel = Relationship(shortPredicate, longPredicate, comparers);
+                            if (shorterRel != RelationshipType.Incomparable)
+                            {
+                                // Found related predicates
+                                break;
+                            }
+                        }
+                        if (shorterRel == RelationshipType.Incomparable)
+                        {
+                            // Predicate in shorter is incomparable so done
+                            soFar = RelationshipType.Incomparable;
+                            break;
+                        }
+                        else
+                        {
+                            if (soFar == RelationshipType.Incomparable)
+                            {
+                                soFar = shorterRel;
+                            }
+                            if (soFar == RelationshipType.Equal)
+                            {
+                                if (shorterRel == RelationshipType.Generalizes
+                                    || (shorterRel == RelationshipType.Specializes && shorterCount == longerCount)
+                                    || shorterRel == RelationshipType.Equal)
+                                {
+                                    soFar = shorterRel;
+                                }
+                                else
+                                {
+                                    break;
+                                }
+                            }
+                            else if (soFar != shorterRel)
+                            {
+                                // Not continued with sub/super so incomparable
+                                break;
+                            }
+                        }
+                    }
+                }
+                if (shorterCount != longerCount)
+                {
+                    switch (soFar)
+                    {
+                        case RelationshipType.Equal:
+                        case RelationshipType.Generalizes: soFar = RelationshipType.Generalizes; break;
+                        default: soFar = RelationshipType.Incomparable; break;
+                    }
+                }
+                soFar = BindingsRelationship(soFar, shorter, longer);
+            }
+            soFar = IgnoreRelationship(soFar, shorter, longer);
+            return swapped ? soFar.Swap() : soFar;
+        }
+
+        private RelationshipType BindingsRelationship(RelationshipType soFar, Clause shorterClause, Clause longerClause)
```

Comment:
Same suggestion about comments on these methods. I understnad the purpose but I imagine it being harder for people with less context. I'd suggest trying to walk through high level concepts with code now that you have all this fresh :)